### ISSUE TITLE
APC and SMES power draw priority take two

### ISSUE THREAD

**Issue by Knightfall5** [VOREStation/VOREStation#2798]
#### Brief description of the issue
SMES units draw power before APC units now, causing APC units to run out of power if there is no extra power in the grid.

#### What you expected to happen
APC units receive power first, then all spare power is sent to SMES units for storage.

#### What actually happened
All power goes to SMES units first for storage, then spare power to APCs for powering the station.

#### Steps to reproduce
Set substation SMES units, wonder why the tram and civilian areas nearby have no power.

#### Code Revision
e1a9a611a8259561240bb0fb973331f9a9eb413b
 
#### Anything else you may wish to add:
Since the last one was closed before any actual discussion occurred, let's try again. Anyone who doubts that it wasn't like this before can ask Aronai, who was on a server running Vore code from halloween. 

Furthermore, I'm an engineering main and I've been playing here since before the substations existed, and since I liked to set all the substations to max input and let them charge up for an hour I think I'd have noticed a couple years ago if the SMES units were charging first. It has always worked like I describe above, until this update: https://github.com/Cameron653/VOREStation-1/commit/a97a57427801b3ac0b2f7d7bac3619e2e0f4d5ea which directly alters the powernet. 

And lastly, ever since this change the Tram and civilian areas have been going dark. A couple days ago, I joined into two different rounds where these areas had zero power but the engine was running, and the substations were hogging all the power.

**Comment by Screemonster** [comment]
Stop maxing the fucking substations if you don't have enough power in the grid to supply all the substations at max.

Set the inputs to like 50kW until there's a pixel worth of charge in them, flip the substation bypass, increase outputs until departments are supplied, increase inputs to slightly higher, then when all departments are supplied, raise substation inputs until the powernet is at capacity. If you don't have enough power, you might want to look into constructing additional generation or building a secondary main SMES since the engine will happily kick out more power than the main SMES will provide.

tl;dr if you want to control where the power goes in the powernet, use the tools that are already available to you instead of just slamming every dial to maximum and hoping for the coders to account for your slapdash approach.

**Comment by SpadesNeil** [comment]
ANGEREY SCREE

**Comment by SpadesNeil** [comment]
Also was the original issue closed or no? I'm tempted to mark this as a duplicate.

**Comment by Knightfall5** [comment]
Instead of inconveniencing the people who set up power every round, why not just fix the issue? And yes, when civilian areas are constantly going dark there is an issue here. Your advice is also stuff I already know about, been here awhile and am engineering main, but I can't and won't set up power every round I am here, that isn't reasonable.

And yes, the original issue was closed a few days ago on the grounds that "It was actually always like this", and I've proven otherwise to a dev.

**Comment by Arokha** [comment]
Title as written is not true, since a simple test of connecting SMES and APC on the same wire to a generator shows the APC charges fine while the SMES charges partially.

It may be that the map's powernet is set up in such a way that favors SMESes over APCs, or that at map load, the powernet is loaded in such an order that favors APCs, but it's not true in every case as simply building a new SMES and APC will show (or even mapping one in, which seems to dispute that it's based on map load order)

**Comment by Screemonster** [comment]
I hardly think "doing their job properly" is being inconvenienced.

**Comment by Knightfall5** [comment]
If "doing their job properly" is a hassle for players, they aren't going to do it properly. Or at all. Source, I actually play engineering and it is a hassle for me. And yes, that simple test you did resulted in the APC charging first as it should, but it doesn't happen like that in the main powernet that we actually use, which is where the issue is.

**Comment by Knightfall5** [comment]
This is a note at the top of that merge I listed : * NOTE: The global machines list is **no longer sorted** for performance reasons.  As far as I know, the only module that actually ever cared was cameras. 

If I had to take an educated guess, I'd say this would explain why APCs can charge before SMES units, and SMES units can charge before APCs.

**Comment by Arokha** [comment]
It was sorted alphabetically, not using any sort of system in particular to allow one thing to charge before others.

**Comment by Cyantime** [comment]
Just posting some findings here while testing the solar shed APC recharging.
With all three solar SMES at full input and output, the shed APC itself does not charge, but the solar farm APC outside does:
![charge1](https://user-images.githubusercontent.com/29438599/35170321-cdc12258-fd2d-11e7-96a8-e50f13e5ffe3.png)

After desconstructing the shed APC and making a new one along the north wall, it charges normally.
![charge](https://user-images.githubusercontent.com/29438599/35170384-04ba2f0c-fd2e-11e7-86f9-7d0b98a3236f.png)

From the order that the solar farm SMES charge,
![charge3](https://user-images.githubusercontent.com/29438599/35170410-29d113f0-fd2e-11e7-9448-9f5cb84471d3.png)
 in addition to the main distro SMES having priority over the engine room SMES, it seems very likely that SMES and APC power are both being delivered in order of north to south, and east to west.

**Comment by Arokha** [comment]
Makes sense. That's the order maps initialize in.
